### PR TITLE
Feature/prevent vote abuse

### LINKED ISSUE #116: Preventing vote abuse
**Is your feature request related to a problem? Please describe.**
As a developer, I wish to prevent a single user from upvoting or downvoting sometimes multiple times with a single account.

**Describe the solution you'd like**
Each user profile only gets one upvote or downvote on a comment or post.

**Describe alternatives you've considered**
To be considered when the issue is being approved.

**Additional context**
Blocked by #92 (Authentication)
Child issue of #182 (Modifying tasks for authentication)

### PR DESCRIPTION
**Pull request by RajanGuptaNZ**
Fixes #116 

## Proposed Changes
- Voting endpoint now uses middleware authentication
- User db now stores which posts a user has voted on
- Users can only cast one vote at a time, if the endpoint is called after they have already voted, their initial vote is removed and a new one is cast
- "upvote" : "false" will now delete their vote regardless of "upvote_type"

### CODE REVIEW

File: server/src/routes/posts.js [hunk 1]
```diff
@@ -123,40 +123,110 @@ router.delete('/:id', firebaseAuthMiddleware, async (req, res) => {
...
+        await Post.updateOne({ _id: req.params.id }, { upvotes_laugh: laughs - 1 });
+        await User.updateOne(
+          { _id: currentUser._id },
+          { $pull: { 'votes.laughs': req.params.id } },
+        );
+      } else if (checkSad) {
+        await Post.updateOne({ _id: req.params.id }, { upvotes_sad: sads - 1 });
+        await User.updateOne({ _id: currentUser._id }, { $pull: { 'votes.sads': req.params.id } });
+      }
+    } else if (checkClap && req.body.upvote_type !== 'clap') {
```
Comment (jacoballen4534) [inline server/src/routes/posts.js:182]: I was under the impression it was just laughing and sad that cancelled each other out (up /down vote) and you could stack a clap on as well

Comment (RajanGuptaNZ) [reply server/src/routes/posts.js:182]: Ah I assumed you could just have one vote max, like fb. If we can confirm the expected behaviour I'll fix it up. @mmir415 @MingChen91

Comment (MingChen91) [reply server/src/routes/posts.js:182]: I think originally you could have multiple, but I think it's fine to have just 1 max ? @jacoballen4534 @RajanGuptaNZ  thoughts?

Comment (jacoballen4534) [reply server/src/routes/posts.js:182]: Guess that makes sense

File: server/src/routes/posts.js [hunk 2]
```diff
@@ -123,40 +123,110 @@ router.delete('/:id', firebaseAuthMiddleware, async (req, res) => {
...
-      );
+  // Check body inputs are valid
+  if (currentPost == null) {
+    res.status(400).json({ message: 'Invalid post ID' });
+    return;
+  }
+  if (currentUser == null) {
+    res.status(400).json({ message: 'Invalid user ID' });
+    return;
+  }
```
Comment (MingChen91) [inline server/src/routes/posts.js:137]: Could you just create a new user object for them if they are not already in the database? Something like
```
if (!currentUser) {
      currentUser = new User({
        email: req.user.email,
      });
  await currentUser.save();
} 
```

Comment (gracecocks) [reply server/src/routes/posts.js:137]: Creating a user if its non-existent is done above in the create post endpoint, so could just be copied from there :)

Comment (RajanGuptaNZ) [reply server/src/routes/posts.js]: Added in latest commit :)

**Comment by MingChen91** [comment]
pulled and tested , works as expected, cheers

**Pull request merged**